src/codegen/barrier.cpp

```diff
@@ -0,0 +1,39 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// barrier.cpp
+//
+// Identification: src/codegen/barrier.cpp
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "common/logger.h"
+#include "codegen/barrier.h"
+#include "codegen/utils/oa_hash_table.h"
+
+namespace peloton {
+namespace codegen {
+
+void Barrier::InitInstance(Barrier *ins, uint64_t n_workers)
+{
+    ins->SetBarrier(new boost::barrier(n_workers));
+    ins->SetWorkerCount(n_workers);
+    ins->InitGlobalHashTableMergeLock();
+}
+
+void Barrier::MasterWait()
+{
+    while(n_workers_ > 0);
```

Comment:
might better considering using thread sleep/yield instead of spinning over variables